examples/apps/attributable-map/tsconfig.json

```diff
@@ -1,10 +1,12 @@
 {
-	"extends": "@fluidframework/build-common/ts-common-config.json",
+	"extends": [
+		"../../../common/build/build-common/tsconfig.base.json",
+		"../../../common/build/build-common/tsconfig.esm-only.json",
+	],
```

Comment:
All of the esm-only packages should set `"type": "module"` in package.json.

Reply:
I guess my comment is kind of wrong because failed to double check what is in `esm-only`. This isn't an ESM build. It is a Node16 CJS build. `tsconfig.esm-only.json` should really be renamed to `tsconfig.node16.json`. Maybe for now create that one (`tsconfig.node16.json`) and then we can transition away from it in subsequent PR.

Reply:
Since this was ESM (module was esnext) before probably want to restore to that with package.json "type": "module".

Reply:
Maybe you have a different naming convention in kind than I do. I'd like to name the configs after the "build kind" of the package. CJS, ESM, and dual build. Is there an advantage to a different scheme that I'm missing?

Reply:
When module is Node16 that doesn't determine whether the build is ESM or CJS. It could be either.
I think it is hard to name things by how you intend to use them. You really have to name them for what they are.

Reply:
The readme is okay. But this change didn't follow the rules and I think that helps illustrate the trouble with the naming. I think anyone looking at this would think the intention was an ESM build. But that isn't what happened. Without any other changes the `lib` directory of this package has CJS files in it.
We maybe could keep that naming if there was something that verified any use of tsconfig.esm-only.json was accompanied by package.json "type": "module".

Reply:
This statement in the readme under tsconfig.esm.json is a little incorrect:

"This config uses Node10 module resolution because Node16 cannot be used to build both CJS and ESM from a common source."

Reply:
> Since this was ESM (module was esnext) before probably want to restore to that with package.json "type": "module".

These are the only valid combinations:

Node16/Node16, nodenext/nodenext, or ESNext/Bundler

> Because node16 and nodenext are the only module options that reflect the complexities of Node.js’s dual module system, they are the only correct module options for all apps and libraries that are intended to run in Node.js v12 or later, whether they use ES modules or not.

https://www.typescriptlang.org/docs/handbook/modules/reference.html#node16-nodenext

> - Use esnext with --moduleResolution bundler for bundlers, Bun, and tsx.
> - Do not use for Node.js. Use node16 or nodenext with "type": "module" in package.json to emit ES modules for Node.js.

I'll update the ESM only packages to have type module tomorrow, now that the config file renames are merged.

Reply:
Those aren't the combinations we are using. Tsc-multi ESM is using ESNext/Node16.
If we are saying CommonJS/Node10 is not a valid combination, then we need to stop the dual build or get package.json type hooking in place. I was able to spend a little more time on that Thursday.

I view the CJS build as there for legacy support; so if it is Node10, I think it is fine. Also, as far as I know CommonJS/Node10 versus type:CommonJs+Node16/Node16 does not emit different code today. At least the test case I had, showed no differences.

Looking forward to "type": "module" :)

Reply:
> Those aren't the combinations we are using. Tsc-multi ESM is using ESNext/Node16.

Oh I see your point.

> I view the CJS build as there for legacy support; so if it is Node10, I think it is fine.

Not once we start adding and using export paths. Node10 doesn't resolve the exports field afaik. Runtime is fine, but compile time tsc doesn't resolve it correctly.

https://www.typescriptlang.org/docs/handbook/modules/reference.html#node10-formerly-known-as-node

Reply:
Re package.json hooking, have you looked at https://github.com/isaacs/tshy ? The approach it takes it exactly what you're proposing, I think, but it is not super configurable from what I can see. Still might be a better piece of infrastructure than tsc-multi for dual-emit with tsc alone.

Reply:
There are two sides to each package. There is the consumption of the package and the production of the package. tsconfig is for the production and doesn't have any play in the consumption apart from needing to get the output right for the consumer. (I know you realize this, but I am writing it out so we are on exactly the same page.)
Saying Node10 in production says we expect to be used by someone that is using Node10. It does not say we can't be used by someone using Node16+. Currently module=CJS+modRes=Node10 produces the same output as type=commonjs+module=Node16[+modRes=Node16] in the cases I have tested. The exports in package.json do not impact production unless we add self-references and generally we don't.
We can have a package with defined exports and .*js code produced any which way.
If the consumer uses Node10 resolution, then they can skip the exports and use main to get to public or whatever we have specified there.